collective-big-count/test_reduce.c

```diff
@@ -0,0 +1,187 @@
+    }
+#endif
+
+    /*
+     * All done
+     */
+    MPI_Finalize();
+    return ret;
+}
+
+int my_c_test_core(MPI_Datatype dtype, size_t total_num_elements, bool blocking)
+{
+    int ret = 0;
+    size_t i;
+    MPI_Request request;
+    char *mpi_function = blocking ? "MPI_Reduce" : "MPI_Ireduce";
+
+    // Actual payload size as divisible by the sizeof(dt)
+    size_t payload_size_actual;
+
+    /*
+     * Initialize vector
+     */
+    int *my_int_recv_vector = NULL;
+    int *my_int_send_vector = NULL;
+    double _Complex *my_dc_recv_vector = NULL;
+    double _Complex *my_dc_send_vector = NULL;
+    size_t num_wrong = 0;
+
+    assert(MPI_INT == dtype || MPI_C_DOUBLE_COMPLEX == dtype);
+    assert(total_num_elements <= INT_MAX);
+
+    if( MPI_INT == dtype ) {
+        payload_size_actual = total_num_elements * sizeof(int);
+        if (world_rank == 0) {
+            my_int_recv_vector = (int*)safe_malloc(payload_size_actual);
+        }
+        my_int_send_vector = (int*)safe_malloc(payload_size_actual);
+    } else {
+        payload_size_actual = total_num_elements * sizeof(double _Complex);
+        if (world_rank == 0) {
+            my_dc_recv_vector = (double _Complex*)safe_malloc(payload_size_actual);
+        }
+        my_dc_send_vector = (double _Complex*)safe_malloc(payload_size_actual);
+    }
+
+    for(i = 0; i < total_num_elements; ++i) {
+        if( MPI_INT == dtype ) {
+            my_int_send_vector[i] = 1;
+            if (world_rank == 0) {
+                my_int_recv_vector[i] = -1;
+            }
+        } else {
+            my_dc_send_vector[i] = 1.0 - 1.0*I;
+            if (world_rank == 0) {
+                my_dc_recv_vector[i] = -1.0 + 1.0*I;
+            }
+        }
+    }
+
+    /*
+     * MPI_Allreduce fails when size of my_int_vector is large
+     */
+    if (world_rank == 0) {
+        printf("---------------------\nResults from %s(%s x %zu = %zu or %s):\n",
+               mpi_function, (MPI_INT == dtype ? "int" : "double _Complex"),
+               total_num_elements, payload_size_actual, human_bytes(payload_size_actual));
+    }
+    if (blocking) {
+        if( MPI_INT == dtype ) {
+            MPI_Reduce(my_int_send_vector, my_int_recv_vector,
+                       (int)total_num_elements, dtype,
+                       MPI_SUM, 0, MPI_COMM_WORLD);
+        } else {
+            MPI_Reduce(my_dc_send_vector, my_dc_recv_vector,
+                       (int)total_num_elements, dtype,
+                       MPI_SUM, 0, MPI_COMM_WORLD);
+        }
+    }
+    else {
+        if( MPI_INT == dtype ) {
+            MPI_Ireduce(my_int_send_vector, my_int_recv_vector,
+                       (int)total_num_elements, dtype,
+                       MPI_SUM, 0, MPI_COMM_WORLD, &request);
+        } else {
+            MPI_Ireduce(my_dc_send_vector, my_dc_recv_vector,
+                       (int)total_num_elements, dtype,
+                       MPI_SUM, 0, MPI_COMM_WORLD, &request);
```

Comment:
For future, could be enhanced to test various MPI_OPs.  FUTURE! :)

Reply:
Yep, lots of good future work items. Good items for folks new to the project and want to write some MPI code. PRs always welcome :)